unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -106,92 +168,125 @@ public int length() {
+  }
+
+  private boolean matchAt(final UTF8String s, int pos) {
+    if (s.numBytes + pos > numBytes || pos < 0) {
       return false;
     }
-    int i = 0;
-    while (i < prefix.length && prefix[i] == bytes[i + offsetInBytes]) {
-      i++;
-    }
-    return i == prefix.length;
+    return ByteArrayMethods.arrayEquals(base, offset + pos, s.base, s.offset, s.numBytes);
   }
 
   public boolean startsWith(final UTF8String prefix) {
-    return startsWith(prefix.getBytes(), 0);
+    return matchAt(prefix, 0);
   }
 
   public boolean endsWith(final UTF8String suffix) {
-    return startsWith(suffix.getBytes(), bytes.length - suffix.getBytes().length);
+    return matchAt(suffix, numBytes - suffix.numBytes);
   }
 
   public UTF8String toUpperCase() {
-    return UTF8String.fromString(toString().toUpperCase());
+    // It's always have the same number of bytes for upper case
+    byte[] buf = new byte[numBytes];
+    copyMemory(base, offset, buf, BYTE_ARRAY_OFFSET, numBytes);
+
+    for (int i = 0; i < numBytes; ){
+      int n = numBytesForFirstByte(getByte(i));
+      int code = codePointAt(i, n);
+      int upper = Character.toUpperCase(code);
+      if (upper != code) {
+        updateCodePoint(buf, BYTE_ARRAY_OFFSET + i, upper, n);
+      }
+      i += n;
+    }
+
+    return fromBytes(buf);
   }
 
   public UTF8String toLowerCase() {
-    return UTF8String.fromString(toString().toLowerCase());
+    // It's always have the same number of bytes for upper case
+    byte[] buf = new byte[numBytes];
+    copyMemory(base, offset, buf, BYTE_ARRAY_OFFSET, numBytes);
+
+    for (int i = 0; i < numBytes; ){
+      int n = numBytesForFirstByte(getByte(i));
+      int code = codePointAt(i, n);
+      int lower = Character.toLowerCase(code);
+      if (lower != code) {
+        updateCodePoint(buf, BYTE_ARRAY_OFFSET + i, lower, n);
+      }
+      i += n;
+    }
+    return fromBytes(buf);
   }
 
   @Override
   public String toString() {
     try {
-      return new String(bytes, "utf-8");
+      // this is slow
+      return new String(getBytes(), "utf-8");
     } catch (UnsupportedEncodingException e) {
       // Turn the exception into unchecked so we can find out about it at runtime, but
       // don't need to add lots of boilerplate code everywhere.
-      PlatformDependent.throwException(e);
+      throwException(e);
       return "unknown";  // we will never reach here.
     }
   }
 
   @Override
   public UTF8String clone() {
-    return new UTF8String().set(bytes);
+    return fromBytes(getBytes());
   }
 
   @Override
   public int compareTo(final UTF8String other) {
-    final byte[] b = other.getBytes();
-    for (int i = 0; i < bytes.length && i < b.length; i++) {
-      int res = bytes[i] - b[i];
+    int len = numBytes < other.numBytes ? numBytes : other.numBytes;
+    int i = 0;
+    // TODO: compare 8 bytes as unsigned long
+    while (i < len) {
+      int res = (getByte(i) & 0xFF) - (other.getByte(i) & 0xFF);
```

Comment:
Maybe we should add a comment here to explain that this is performing unsigned comparison.